pkg/vm/engine/tae/containers/cn_vector_adapter.go

```diff
@@ -0,0 +1,644 @@
+
+	err = vec.downstreamVector.UnmarshalBinary(downStreamVectorByteArr)
+
+	return
+}
+
+func (vec *CnTaeVector[T]) CloneWindow(offset, length int, allocator ...*mpool.MPool) Vector {
+	opts := Options{}
+	if len(allocator) == 0 {
+		opts.Allocator = vec.GetAllocator()
+	} else {
+		opts.Allocator = allocator[0]
+	}
+
+	// Create a new NewCnTaeVector
+	cloned := NewCnTaeVector[T](vec.GetType(), vec.Nullable(), opts)
+
+	// Note: We are using ForeachWindow instead of cnVector.Window(vecDup, offset, offset+length, cloned.downstreamVector)
+	// Reason: Using cnVector.Window(...) is giving "panic: internal error: mp header corruption"
+	//		   in tables_test.go#TestTxn6(...), when vec.Close() is invoked.
+	// 		   Attaching the cnVector.Window(...) code for reference.
+	// My thoughts: 1. I think, the issue was because cnVector.Window(...)
+	//				- is allocated by mpool1 (the allocator... argument)
+	//				- and Free() by mpool2, (the vec.mpool field)
+	//
+	// 			    2. I did try only allocating using mpool2 alone, but that didn't work.
+
+	/*
+		cloned.isAllocatedFromMpool = true
+
+		// Create a duplicate of the downstream CN vector
+		vecDup, _ := cnVector.Dup(vec.downstreamVector, opts.Allocator)
+
+		// Attach that downstream duplicate to the window and perform window action.
+		// The result is subset of downstream vector.
+		cloned.downstreamVector = cnVector.Window(vecDup, offset, offset+length, cloned.downstreamVector)
+	*/
+
+	op := func(v any, _ int) error {
+		cloned.Append(v)
+		return nil
+	}
+	err := vec.ForeachWindow(offset, length, op, nil)
+	if err != nil {
+		return nil
+	}
+
+	return cloned
+}
+
+func (vec *CnTaeVector[T]) Window(offset, length int) Vector {
+
+	// In DN Vector, we are using SharedReference for Window.
+	// In CN Vector, we are creating a new Clone for Window.
+	// So inorder to retain the nature of DN vector, we had use CnTaeVectorWindow Adapter.
+	return &CnTaeVectorWindow[T]{
+		ref: vec,
+		windowBase: &windowBase{
+			offset: offset,
+			length: length,
+		},
+	}
+}
+
+// TODO: --- We can remove below functions as they don't have any usage
+
+func (vec *CnTaeVector[T]) IsView() bool {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) GetView() VectorView {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) DataWindow(offset, length int) []byte {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) Data() []byte {
+	panic("Soon Deprecated")
+}
+
+func (vec *CnTaeVector[T]) SlicePtr() unsafe.Pointer {
+	slice := vec.Slice().([]T)
+	return unsafe.Pointer(&slice[0])
+}
+
+func (vec *CnTaeVector[T]) AppendNoNulls(s any) {
+	panic("Soon Deprecated")
+}
+
+// TODO: --- We can remove below function as they are only used in Testcases.
+
+func (vec *CnTaeVector[T]) Delete(delRowId int) {
+	deletes := roaring.BitmapOf(uint32(delRowId))
+	vec.Compact(deletes)
+}
+
+func (vec *CnTaeVector[T]) ReadFromFile(f common.IVFile, buffer *bytes.Buffer) (err error) {
```

Comment:
It can be removed after CodecTool being removed

Reply:
sure.